// chillax/dist/parallax-utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getPageScroll() {
    var doc = document.documentElement;
    var top = (window.pageYOffset || doc.scrollTop) - (doc.clientTop || 0);
    return top;
}
exports.getPageScroll = getPageScroll;
function clamp(value, min, max) {
    return Math.min(Math.max(value, min), max);
}
exports.clamp = clamp;
function makeArray(arraylike) {
    return Array.prototype.slice.call(arraylike);
}
exports.makeArray = makeArray;
function getViewportHeight() {
    return window.innerHeight;
}
exports.getViewportHeight = getViewportHeight;
function getTopOffsetRelativeToPage(element) {
    var top = 0;
    do {
        top += element.offsetTop || 0;
        element = element.offsetParent;
    } while (element);
    return top;
}
exports.getTopOffsetRelativeToPage = getTopOffsetRelativeToPage;
function getScrollProgressThroughElement(_a) {
    var scroll = _a.scroll, viewportHeight = _a.viewportHeight, element = _a.element;
    var top = getTopOffsetRelativeToPage(element);
    var height = element.offsetHeight;
    var bottom = top + height;
    var progressTop = top - viewportHeight;
    var progressHeight = height + viewportHeight;
    var progress = (scroll - progressTop) / progressHeight;
    return clamp(progress, 0, 1);
}
exports.getScrollProgressThroughElement = getScrollProgressThroughElement;
